// bloomer/esm/grid/grid.js
import * as tslib_1 from "tslib";
import { isDesktop, isFullHD, isMobile, isTablet, isTouch, isWidescreen } from './../bulma';
import { isBetween, isOption } from './../helpers';
var isValidSize = isBetween(1, 12);
var isPlatform = isOption(isMobile, isTablet, isTouch, isDesktop, isWidescreen, isFullHD);
export function getGridSizesModifiers(isSize, isOffset, platform) {
    if (isOffset === void 0) { isOffset = false; }
    if (platform === void 0) { platform = false; }
    var _a;
    return (Number.isInteger(isSize) && isValidSize(isSize)) ? (_a = {},
        _a["is-" + (isOffset ? 'offset-' : '') + isSize + (isPlatform(platform) ? "-" + platform : '')] = true,
        _a) : {};
}
export function removeGridSizesProps(props) {
    var isSize = props.isSize, rest = tslib_1.__rest(props, ["isSize"]);
    return rest;
}
var fractions = {
    '1/2': 'half',
    '1/3': 'one-third',
    '1/4': 'one-quarter',
    '2/3': 'two-thirds',
    '3/4': 'three-quarters',
};
var width = {
    full: 'full',
    narrow: 'narrow',
};
function getGridFractionsModifiers(size, isOffset, platform) {
    var _a;
    var sizes = !isOffset ? tslib_1.__assign({}, fractions, width) : fractions;
    return sizes[size] ? (_a = {}, _a["is-" + (isOffset ? 'offset-' : '') + sizes[size] + (isPlatform(platform) ? '-' + platform : '')] = true, _a) : {};
}
function getGridObjectSizeModifiers(size, isOffset) {
    return Object.keys(size).reduce(function (acc, key) {
        if (key === 'default')
            return tslib_1.__assign({}, acc, getHorizontalSizeModifiers(size[key], isOffset));
        return isPlatform(key) ? tslib_1.__assign({}, acc, getHorizontalSizeModifiers(size[key], isOffset, key)) : acc;
    }, {});
}
function getHorizontalSizeModifiers(size, isOffset, platform) {
    if (isOffset === void 0) { isOffset = false; }
    if (platform === void 0) { platform = false; }
    if (typeof size === 'number')
        return getGridSizesModifiers(size, isOffset, platform);
    if (typeof size === 'string')
        return getGridFractionsModifiers(size, isOffset, platform);
    if (typeof size === 'object')
        return getGridObjectSizeModifiers(size, isOffset);
    return {};
}
export function getSizeModifiers(props) {
    return getHorizontalSizeModifiers(props.isSize);
}
export function removeSizeProps(props) {
    return removeGridSizesProps(props);
}
export function getOffsetModifiers(props) {
    return getHorizontalSizeModifiers(props.isOffset, true);
}
export function removeOffsetProps(props) {
    var isOffset = props.isOffset, rest = tslib_1.__rest(props, ["isOffset"]);
    return rest;
}
