// react-overlay-loader/lib/Loader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require("react");

var React = _interopRequireWildcard(_react);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

var styles = {
  text: {
    marginTop: 52,
    color: "#888",
    marginLeft: 6
  },
  spinner: {
    backgroundColor: "rgba(255, 255, 255, 0.5)",
    width: "100%",
    height: "100%",
    top: 0,
    left: 0,
    zIndex: 1000,
    display: "flex",
    alignItems: "center",
    justifyContent: "center"
  }
};

var Loader = function Loader(_ref) {
  var loading = _ref.loading,
      _ref$text = _ref.text,
      text = _ref$text === undefined ? "Loading.." : _ref$text,
      fullPage = _ref.fullPage,
      containerStyle = _ref.containerStyle,
      textStyle = _ref.textStyle;

  if (!loading) {
    return null;
  }
  var mergedContainerStyle = _extends({}, styles.spinner, containerStyle, {
    position: fullPage ? "fixed" : "absolute"
  });
  return React.createElement(
    "div",
    { style: mergedContainerStyle },
    React.createElement("div", { className: "react-overlay-loader-spinner" }),
    React.createElement(
      "span",
      { style: _extends({}, styles.text, textStyle) },
      text
    )
  );
};

exports.default = Loader;
